### PR TITLE
OMP Collapse 

### PR DESCRIPTION
**Pull request by artv3**
Code collapses for 2 or 3 for loops using the #pragma omp collapse clause.

### CODE REVIEW

**Comment by rhornung67** [comment]
@artv3 This looks basically like we would want it to work (we'll have others weigh in on the naming convention of collapse policy). I don't understand why you don't have execution policies in the 'For' definitions. What is this assuming about those.

Have you tried this with a triply-nested loop as follows?

using Pol = RAJA::nested::Policy<RAJA::nested::ompCollapse<
                    RAJA::nested::For<0>,
                    RAJA::nested::For<1> >, 
                    RAJA::nested::For<2, RAJA::seq_exec> >
i.e., collapse the outer two loops and run a sequential for-loop for each thread as an inner loop.

Or, by putting a sequential loop on the outside and collapsing the inner two?

**Comment by artv3** [comment]
@rhornung67, I was trying to get naive working version first. I chatted with @MrBurmark and @ajkunen about the code but we realized that we have a race condition when assigning indices to threads. We have not yet found a solution for it....

**Comment by rhornung67** [comment]
@artv3 I get it. I was slightly confused (I can't keep up with everything going on -- sorry). I spoke to @ajkunen -- I'm on the same page as you guys now.  Keep up the good work!

**Comment by ajkunen** [comment]
@artv3  I had to add a thread_privatize that would work for ompCollapse.  Take a look, it seems to work.  You should be able to implement deeper collapsings with it.

**Comment by artv3** [comment]
Thanks @ajkunen, unfortunately I seem to be getting segfaults when I try it on the Jacobi example. I haven't been able to pinpoint the cause.

**Comment by artv3** [comment]
@rhornung67, @ajkunen . It turns out the code does work, just not with GCC 4.9 with optimizations turned on.... I found success with Clang 3.9 and Intel 17.0. GCC 4.9 does not properly copy the lambda to the non-master threads which have zeros for the captured variables.  Thanks @MrBurmark !

**Comment by rhornung67** [comment]
@artv3 @ajkunen it looks like some of the nested tests are not compiling since the compilers think that loops are not perfectly nested. Any idea what's up with that?

**Comment by artv3** [comment]
@rhornung67 , I was observing segfaults indeed. But it was because the lambdas were failing to capture the pointers, the debugger showed they were NULL. It was only happening with GCC though, I didn't have a problem with intel or clang.

**Comment by ajkunen** [comment]
@artv3 I think I've fixed the segfaults, we just needed to hold on to the thread privatizer object which holds the LoopData object.  the object in get_priv() just contains a reference to the privatizer.  You can thank @trws for this lovely pattern :)

**Comment by artv3** [comment]
Thanks @ajkunen ! Intel compiler still giving problems though. Almost there though.

**Comment by artv3** [comment]
@rhornung67 , @ajkunen, I was able to fix the bug with some pointers from @jonesholger and @MrBurmark . Regarding the loops such as 

using Pol = RAJA::nested::Policy<RAJA::nested::ompCollapse<
RAJA::nested::For<0>,
RAJA::nested::For<1> >,
RAJA::nested::For<2, RAJA::seq_exec> >
i.e., collapse the outer two loops and run a sequential for-loop for each thread as an inner loop.

We can't do this yet, there is some issues with camp and counting policies and range segments. Perhaps @ajkunen or @trws could provide some pointers? I just need to add some unit test.

File: include/RAJA/policy/openmp/nested.hpp [hunk 1]
```diff
@@ -130,6 +134,66 @@ namespace nested
...
+    void operator()(Collapse<omp_parallel_collapse_exec, FT0, FT1, FT2> const &, WrappedBody const &wrap)
+    {
+
+      auto b0 = std::begin(camp::get<FT0::index_val>(wrap.data.st));
+      auto b1 = std::begin(camp::get<FT1::index_val>(wrap.data.st));
+      auto b2 = std::begin(camp::get<FT2::index_val>(wrap.data.st));
+      
+      auto e0 = std::end(camp::get<FT0::index_val>(wrap.data.st));
+      auto e1 = std::end(camp::get<FT1::index_val>(wrap.data.st));
+      auto e2 = std::begin(camp::get<FT2::index_val>(wrap.data.st));
```
Comment (MrBurmark) [inline include/RAJA/policy/openmp/nested.hpp]: Should be end.

Comment (rhornung67) [reply include/RAJA/policy/openmp/nested.hpp]: Good catch!

**Comment by artv3** [comment]
Thanks @MrBurmark . @rhornung67 , I am working on the omp target version, but I believe 
private_wrap.data.template is causing run-time issues. I was wondering if I need to do anything else to make it omp target compatible? @trws @DavidPoliakoff

**Comment by artv3** [comment]
@rhornung67 @ajkunen , I added a test in nested.cpp which breaks with gnu optimizations.

**Comment by artv3** [comment]
Works on my end now. Thanks @ajkunen.

**Comment by ajkunen** [comment]
looks like gnu compiler craps out when combining nowait and collapse... but the use of the parallel region around collapse makes the nowait unnecessary, so I've removed it.  Everything looks good now.

**Comment by artv3** [comment]
Thanks @ajkunen . Weird bug.... Happy Holidays everyone!

File: include/RAJA/policy/openmp/nested.hpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+ * \brief   RAJA header file containing constructs used to run forallN
+ *
+ ******************************************************************************
+ */
+
+#ifndef RAJA_policy_openmp_nested_HPP
+#define RAJA_policy_openmp_nested_HPP
+
+#include "RAJA/config.hpp"
+#include "RAJA/pattern/nested.hpp"
```
Comment (rhornung67) [inline include/RAJA/policy/openmp/nested.hpp]: @artv3 please move header inclusions, other than RAJA/config.hpp, inside the RAJA_ENABLE_OPENMP guard.  Also, replace the copyright stuff with the short form we are now using (see openmp/forall.hpp to see what I mean).

File: include/RAJA/policy/openmp/nested.hpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+#include <cassert>
+#include <climits>
+
+#include "RAJA/RAJA.hpp"
+#include "RAJA/config.hpp"
+#include "RAJA/util/defines.hpp"
+#include "RAJA/util/types.hpp"
+
+//#include "RAJA/policy/openmp/MemUtils_OPENMP.hpp"
```
Comment (rhornung67) [inline include/RAJA/policy/openmp/nested.hpp]: Please remove the commented out header file include lines if they are not needed.

**Pull request merged**